Fix #5473: Add wiki guide for upgrading the app's compile/target SDK versions

BenHenning opened the pull request:
## Explanation

Fixes #5473

This PR introduces a new wiki page that describes the process by which we can carefully upgrade the app to a newer compile/target SDK version (with a specific emphasis on the target SDK version).

This is largely meant to capture past processes we have followed since those weren't documented in an easily reproducible way. The intent is for this now-documented process to be trialed by upgrading the app to SDK 34 in order to address #5535 (though that technically means that part (1) of the documented process has already been completed).

This PR does not:
- Address exactly when the compile version should be updated, except when required by the target SDK version. It's expected that the team will perform an out-of-band analysis for cases when the compile SDK actually needs to be upgraded sooner than the target SDK.
- Specify a strongly enforced cadence for upgrading the target SDK since the Google Play Console and Google Play app policy seems to already perform this function, though it does add a recommendation for checking for updates periodically (July/August every year).

## Essential Checklist
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
N/A -- This is introducing a new wiki page and doesn't change the app itself.

BenHenning committed:
Add guide for upgrading the compile/target sdks.

BenHenning committed:
Small follow-up fixes.

BenHenning reviewed (commented):
Self-reviewed changes (brief skim).

BenHenning commented:
PTAL @adhiamboperes for code owners.

PTAL @Rd4dev for just general thoughts on whether the wiki page is well-organized and well-written, and if you have any follow-up suggestions. Please feel free to consider this review optional. If you have the time, I'd appreciate any thoughts you have on this page.

Rd4dev commented on wiki/Upgrading-Target-Sdk-Guide.md:
```diff
@@ -0,0 +1,148 @@
+    B(Part 2: Auditing the Android OS changelog) -->|Audit complete| C
+    C(Part 3: Testing the app & filing problems) -->|Testing finished & issues filed| D
+    D(Part 4: Fixing and stabilizing support) -->|All issues fixed| E
+    E(Part 5: Submitting the upgrade) -->|Compile/target SDK PR submitted| F
```
It looks like the illustration content might be getting cut off due to the text length. Could using a `<br>` tag help with this? (or other approach) 

```mermaid
flowchart TD
  E(Part 5: Submitting the upgrade) -->|Compile/target SDK PR submitted| F(Block)
```

with `<br>` tag:

```mermaid
flowchart TD
  E(Part 5: Submitting<br>the upgrade) -->|Compile/target SDK<br>PR submitted| F(Block)
```

BenHenning replied:
What exactly is being cut off for you?

Here's what I see:

![image](https://github.com/user-attachments/assets/cfc3c11a-f902-417b-8c4e-797aa159d1f4)

Rd4dev replied:
I viewed the page using View File and it renders as below:

![image](https://github.com/user-attachments/assets/c07ebfd8-b8be-4d7e-bf58-86da73af8fc9)
 
Am I viewing it correctly?

Rd4dev replied:
ah, maybe due to varied zoom?

https://github.com/user-attachments/assets/e74f2c87-c4d3-4ef6-b988-fe8f396cbf16

BenHenning replied:
This is actually really interesting. What browser and browser version are you using? This might be a browser-specific GitHub and/or Mermaid bug as zoom doesn't affect things at all for me. From inspecting the graph, it's a generated SVG being rendered.

I'm using Chrome 129.

Rd4dev replied:
I initially tested it on Edge 128, but now I also tested it with
- Chrome 120
- Edge 128
- Chrome 129 (Updated)

seems like on my end, they were still reproducible.

https://github.com/user-attachments/assets/967ae445-7159-4abf-86b3-9039787f6964

Rd4dev replied:
Could any of the OS specific rendering make these differences, I tested it by installing Mint on VM and resembles as you mentioned but the same with Windows 11 renders it differently. Also, this issue - https://github.com/typora/typora-issues/issues/3291#issuecomment-680117090 seems to mention a similar issue for MacOs.

https://github.com/user-attachments/assets/e9c581cb-29c6-4012-822f-506cb1df96fe

# 

But Firefox renders it properly on both VM Mint OS and Windows 11 (which might suggest the earlier observations are less relevant.).

BenHenning replied:
I'm not 100% sure honestly. Browser rendering can get very complicated, so I can only guess on potential compatibility. I tried searching the Mermaid GitHub repo for similar issues and didn't find a perfect match (though there were other text cut-off issues for separate reasons that were addressed). The issue you linked is much closer to what we're observing here than anything I found.

I'm not opposed to adding breaks, but I have little confidence that will fix the problem correctly since a separate display situations might still result in it being cut off. Or, put another way, how much text can we actually fit on a line before it's likely to be cut off? This is further complicated by us not using a monospaced font so line width won't be exactly correlated to number of characters.

Do you have any suggestions on what we could do here? I'm wondering if a different diagram format or maybe just trying to shorten the text would be best (though I thought of that and couldn't come up with good alternatives at the time).

Rd4dev replied:
I tried applying the `text-wrap` property to the node, and it appears to be working. Could this be a viable / consistent workaround?

Tested on few browsers and added them below, could you PTAL!

Snippet:

```
classDef textWrap text-wrap: wrap;

  A(Part 1: With Text Wrap property set to the node):::textWrap -->|Tracking issue exists| B
  B(Part 1: Without Text Wrap property set to the node) -->|Tracking issue exists| C
```

>Tested on Edge 128 

https://github.com/user-attachments/assets/6147dcd5-6824-4940-9747-aea832323d2c


<details open><summary>Chrome 129</summary>

![image](https://github.com/user-attachments/assets/cf2d1fd5-d790-48db-8b8e-895e82dd9792)

</details>

<details><summary>Edge128</summary>

![image](https://github.com/user-attachments/assets/e973dd0b-ea5e-4290-addc-1d485439d1a6)

</details>

<details><summary>Firefox 130</summary>

![image](https://github.com/user-attachments/assets/486aa1eb-02ea-473e-8728-b2762e5d5143)

</details>

<details><summary>Brave 1.70</summary>

![image](https://github.com/user-attachments/assets/5dd40a8c-b575-4221-ab7f-849b3b52e623)

</details>

adhiamboperes replied:
Just mentioning that I have not experienced any cutoff issues on chrome or safari even when zooming.

BenHenning replied:
Very nice suggestion! Applied the suggested changes. PTAL @Rd4dev.

Rd4dev replied:
Thanks, @BenHenning.

Rd4dev commented on wiki/Upgrading-Target-Sdk-Guide.md:
```diff
@@ -0,0 +1,148 @@
+2. A new version of Android has released to users and we perform a periodic check for compatibility (usually around July/August). We may decide to upgrade even if there isn't a mandate.
+3. A feature requires a newer version of Android (note this is unlikely since we generally want to design features to work for all of our users).
+
+If it's deemed that there's a new SDK version to target and the app isn't yet targeting it, a new feature request should be filed (similar to #5535) as long as there isn't an existing tracking issue for this work.
```
>(similar to #5535)

Would it be preferable to include direct links than issue numbers, as MD files don’t seem to link them directly?

BenHenning replied:
I thought about doing that, and it's a good suggestion. Done everywhere on the page.

Rd4dev replied:
Thanks.

Rd4dev commented on wiki/Upgrading-Target-Sdk-Guide.md:
```diff
@@ -0,0 +1,148 @@
+
+Any concerning changes or functionality that could be beneficial to Oppia Android should be noted in the tracking issue for the SDK upgrade (see https://github.com/oppia/oppia-android/issues/5137#issuecomment-1815241974 for a good example of this).
+
+Note that narrowing down these categories isn't a process that can be easily described as a set of steps since new OS features may not even be predictable ahead of time. However, here are some tips that might help:
```
Would splitting them into separate subheadings, for eg. "High-Level Considerations" and "Post Findings," improve structure and enhance readability? - [sample md](https://github.com/Rd4dev/Oppia-Android-Fork-from-Fork/blob/Rd4dev-patch-3/wiki/up.md#part-2-auditing-the-android-os-changelog)

BenHenning replied:
Added new sections as suggested, though I went with 'tips for auditing' and kept the actual high-level instructions in the top-level section.

Rd4dev replied:
Thanks.

Rd4dev commented on wiki/Upgrading-Target-Sdk-Guide.md:
```diff
@@ -0,0 +1,148 @@
+Any concerning changes or functionality that could be beneficial to Oppia Android should be noted in the tracking issue for the SDK upgrade (see https://github.com/oppia/oppia-android/issues/5137#issuecomment-1815241974 for a good example of this).
+
+Note that narrowing down these categories isn't a process that can be easily described as a set of steps since new OS features may not even be predictable ahead of time. However, here are some tips that might help:
+- Look for changes in permissions. This could either be an old permission that's now more restricted, or existing SDK functionality that's now blocked by a new permission (both have occurred in past Android OS updates). If the functionalty and/or permission relates to Oppia Android, it should be noted. Note that some things may be tied to permissions and access control that's not obvious such as:
```
Nit: Spelling of functionality
```suggestion
- Look for changes in permissions. This could either be an old permission that's now more restricted, or existing SDK functionality that's now blocked by a new permission (both have occurred in past Android OS updates). If the functionality and/or permission relates to Oppia Android, it should be noted. Note that some things may be tied to permissions and access control that's not obvious such as:
```

BenHenning replied:
Nice catch! Fixed.

Rd4dev commented on wiki/Upgrading-Target-Sdk-Guide.md:
```diff
@@ -0,0 +1,148 @@
+### Part 3: Testing the app & filing problems
+
+Ensuring compatibility with a target SDK upgrade requires two things:
+1. A local Bazel production build of the app (``//:oppia_beta``) with the target SDK temporarily upgraded (see part (5) below) but not checked in.
```
Would it be helpful to include somewhere a link to the [Bazel Build instructions wiki](https://github.com/oppia/oppia-android/wiki/Oppia-Bazel-Setup-Instructions) here for building the local Bazel production app, or can we assume developers are already familiar with the process?

BenHenning replied:
Linking to the Bazel documentation seems reasonable--done.

Rd4dev replied:
Thanks!

Rd4dev commented on wiki/Upgrading-Target-Sdk-Guide.md:
```diff
@@ -0,0 +1,148 @@
+- In some cases, a real device may need to be used instead of an emulator since certain features change behavior on an emulator (such as the drag and drop interaction).
+- If the app isn't already using the new version of Android as its compile SDK version, then it may fail to build. Any build failures should be filed as issues on the issue tracker and fixed before this part of the process can continue.
+
+The local version of the app should be deployed to an emulator (or a real device if one is available) and tested as such:
```
Would it be beneficial to add subheadings such as 
- Deployment 
- User Behavior Testing 
- General Analysis
- Filing Issues

[sample md](https://github.com/Rd4dev/Oppia-Android-Fork-from-Fork/blob/Rd4dev-patch-3/wiki/up.md#part-3-testing-the-app--filing-problems)

or would it be preferable to maintain the current single flow?

BenHenning replied:
I'm not entirely sold on the specific organizational approach suggested, however I do agree that the overall flow of this section is difficult to follow and, being it's probably the most important section, needed a slight rewrite. I've attempted to reorganize it to have some sub-structure, and to have less redundancy. Is it clearer now or do you think there are additional changes that should be made?

Rd4dev replied:
Thanks for adding the sub sections, those are helpful!

Rd4dev commented on wiki/Upgrading-Target-Sdk-Guide.md:
```diff
@@ -0,0 +1,148 @@
+
+### Part 4: Fixing and stabilizing support
+
+Work on fixing all identified problems from parts (3) and (4) (either by directly fixing the problems via code changes, or via coordination with other members of the team).
```
Same clarification as above, if adding subheadings for better structure would be helpful (may be)? 

- Address Identified Problems
- Verification of Fixes
- Final Confirmation

[sample md](https://github.com/Rd4dev/Oppia-Android-Fork-from-Fork/blob/Rd4dev-patch-3/wiki/up.md#part-4-fixing-and-stabilizing-support)

BenHenning replied:
Since this is a very short section, I'm not sure the subheadings really add much value. Do you find the current section process difficult to follow such that additional organizing might help?

Rd4dev replied:
Yeah, makes sense. The current version by itself is easy to follow, so I'll resolve it. Thanks!

Rd4dev commented on wiki/Upgrading-Target-Sdk-Guide.md:
```diff
@@ -0,0 +1,148 @@
+- Updating the compile-time SDK version may require updating the build tools version. This is **not** a simple change and may cause difficult-to-fix breakages due to subtle compatibility issues between third-party dependencies and the build system configurations. If you suspect a build tools version upgrade is needed, please file an issue to track it and contact the developer workflow team lead t ...
+- Updating tests to use a newer version of Android can be exceptionally complicated, and thus this is considered a completely separate exercise from upgrading production code. See part (6) below for more specifics.
+
+#### Part 5.1: Updating the compiled SDK version
```
I'm not sure if this is needed or if I’m following the structure correctly, but even though Part 3 (Testing) refers to Part 5 (Upgrading), would it be useful to restructure and include a **'How to Upgrade'** section before the testing section (Part 3).

As I believe testing typically follows the upgrade process, providing the 'How to Upgrade' section earlier might create a clearer path. 

1. Identifying the need to upgrade
2. Auditing the Android OS changelog
3. **How to upgrade**
4. Testing the app & filing problems
5. Fixing and stabilizing support
6. Submitting the upgrade
7. Future work items and upgrading Robolectric

Or is the current structure designed to ensure that all necessary testing procedures are addressed before proceeding with the upgrade?

BenHenning replied:
The goal is the latter--we don't actually want to create and submit an upgrade PR until it's time for that to be merged (which is when auditing and testing has completed). I was also a bit concerned with this organizational approach since it leads to lots of jumping around. Some possible options we have:
- Keep it as-is since the overall process is fairly short.
- Introduce an appendix with sections detailing specific tasks that need to be done and how, and update the process to point to these steps (this keeps the process itself linear but still results in jumping to/from the appendix, maybe even more than the current writing).
- Split this into two sections as you have it (though my concern with that is 'how to upgrade' doesn't actually have as strong of a completion indication compared to the other sections).

I slightly lean toward (1) since the other two _seem_ more complicated, but I'm not sold on any of the approaches. Do you have any thoughts?

Rd4dev replied:
Makes sense. I was considering whether that might create a gap in the flow..., but as you pointed out, sticking with option 1 seems like the better approach (for now).

Rd4dev commented on wiki/Upgrading-Target-Sdk-Guide.md:
```diff
@@ -0,0 +1,148 @@
+2. Bazel [``build_vars.bzl``](https://github.com/oppia/oppia-android/blob/dfb9a301280b9a46526cb2f5ca6329532fec6bf0/build_vars.bzl#L1) changes to ``BUILD_SDK_VERSION``.
+3. Gradle ``compileSdkVersion`` changes (e.g. for [``app/build.gradle``](https://github.com/oppia/oppia-android/blob/dfb9a301280b9a46526cb2f5ca6329532fec6bf0/app/build.gradle#L8)). Note that all module ``.gradle`` files will need to be updated in this way.
+
+#### Part 5.2: Updating the target SDK version
```
Would the WORKSPACE file also require an update in this case? (apologies if I overlooked)

BenHenning replied:
Good question--it shouldn't anymore now that we've moved relevant build configurations to build_vars.bzl. It does still need to be updated for the target SDK, but that'll go away soonish (once we remove the APKs targets--they are rarely used and we should prefer using the AAB builds even for local development).

Rd4dev replied:
Ah, thanks for the clarification!

Rd4dev reviewed (commented):
Thanks @BenHenning, would just like to understand two minor things:

1. Would it be useful to restructure/include a 'How to Upgrade' section?
2. Would the WORKSPACE file need an update?

BenHenning committed:
Address review comments.

Also, rename 'part' to 'step' for hopefully better clarity.

BenHenning committed:
Revert accidental typo.

BenHenning reviewed (commented):
Self-reviewed latest changes.

BenHenning commented:
Thanks @Rd4dev! I found your comments very helpful! Could you PTAL at my follow-ups and please let me know if you think any additional changes would be useful?

Rd4dev commented on wiki/Upgrading-Target-Sdk-Guide.md:
```diff
@@ -0,0 +1,155 @@
+
+### Step 4: Fixing and stabilizing support
+
+Work on fixing all identified problems from parts (3) and (4) (either by directly fixing the problems via code changes, or via coordination with other members of the team).
```
can these too be updated as steps?

```suggestion
Work on fixing all identified problems from steps(3) and (4) (either by directly fixing the problems via code changes, or via coordination with other members of the team).
```

BenHenning replied:
Good catch! Fixed.

Rd4dev replied:
Thanks.

Rd4dev reviewed (commented):
Thanks @BenHenning for adding the subsections, and reg. the 'How to upgrade' section, as you mentioned it might be better to stick with the original approach for now and revisit later if needed. 

The graphs seem to render differently at various zoom levels. Please resolve any comments if they're blocking the merge.

BenHenning committed:
Update Upgrading-Target-Sdk-Guide.md

Addressed review comment.

BenHenning reviewed (commented):
Checked latest changes.

BenHenning reviewed (commented):
Thanks @Rd4dev! I followed up to both comments--PTAL.

Rd4dev commented:
Thanks, @BenHenning! The only thing left is the zoom levels. I've attached a video reference as I still see the zoom level affecting them. Can you PTAL!

adhiamboperes reviewed (approved):
Thanks @BenHenning! I think this guide is thorough and well structured. LGTM!

BenHenning committed:
Update Upgrading-Target-Sdk-Guide.md

Update Mermaid chart per RD's suggestion to support text wrapping for better screen size compatibility.

BenHenning reviewed (commented):
Self-reviewed latest changes.

BenHenning reviewed (commented):
Thanks @Rd4dev and @adhiamboperes!

@Rd4dev I think I addressed your open comment--can you please confirm?

Rd4dev commented:
Thanks @BenHenning for the update. I’ve resolved all my open comments.

BenHenning committed:
Merge branch 'develop' into add-guide-for-upgrading-compile-target-sdk-versions

BenHenning commented:
Thanks @Rd4dev!
